tests/WindowsUpdateCheckTest.php

```diff
@@ -20,18 +20,91 @@ public function setUp(): void
     $this->assertNotEmpty($json);
     $data = json_decode($json);
     $this->assertNotNull($data);
 
     // This will throw an exception if it does not pass
     $this->schema->in($data);
 
-    // TODO: add some more realism!
+    return $json;
+  }
+
+  private function getJson($tier, $appVersion, $packages, $isUpdate, $isManual, $currentTime = null)
+  {
+    return json_decode($this->getJsonString($tier, $appVersion, $packages, $isUpdate, $isManual, $currentTime));
+  }
+
+  public function testSimpleResultValidatesAgainstSchema(): void
+  {
+    $json = $this->getJsonString('alpha', '14.0.100', ['khmer_angkor'=>'1.0.2', 'foo'=>'0.1'], 0, 1);
     $this->assertJsonStringEqualsJsonFile(__DIR__ . '/fixtures/WindowsUpdateCheck.14.0.json', $json);
   }
+
+  public function testAlphaStaysInAlpha(): void
+  {
+    $json = $this->getJson('alpha', '13.0.100', [], 0, 1);
+    $this->assertEquals('alpha', $json->bundle->stability);
+    $this->assertEquals('14.0.102.0', $json->bundle->version);
+  }
+
+  public function testBetaUpgradesToStable(): void
+  {
+    $json = $this->getJson('beta', '13.0.77', [], 0, 1);
+    $this->assertEquals('stable', $json->bundle->stability);
+    $this->assertEquals('13.0.109.0', $json->bundle->version);
+  }
+
+  public function testStableStaysInStable(): void
+  {
+    $json = $this->getJson('stable', '12.0.10', [], 0, 1);
+    $this->assertEquals('stable', $json->bundle->stability);
+    $this->assertEquals('13.0.109.0', $json->bundle->version);
+  }
+
+  public function testStableGetsDelayedRollout(): void
+  {
+    // test 1 day after release
+    // from fixture, 13.0.109.0 was released 2020-06-05
+    // rollout schedule is defined in ReleaseSchedule and is based on minute of the hour
+    // mktime is h:m:s, m/d/y (yuk!)
+
+    // out-of-schedule
+    $json = $this->getJson('stable', '12.0.10', [], 0, 0, mktime(0, 20, 0, 6, 6, 2020));
+    $this->assertFalse($json->bundle, 'should not be offered release');
+
+    // in-schedule
+    $json = $this->getJson('stable', '12.0.10', [], 0, 0, mktime(0, 1, 0, 6, 6, 2020));
+    $this->assertEquals('13.0.109.0', $json->bundle->version, 'should be offered release');
+
+    // minor update, out-of-schedule
+    $json = $this->getJson('stable', '13.0.107.0', [], 0, 0, mktime(0, 20, 0, 6, 6, 2020));
+    $this->assertEquals('13.0.109.0', $json->bundle->version, 'should be offered release');
+
+    // test 30 days after release
+
+    // in-schedule, 59 minutes after hour
+    $json = $this->getJson('stable', '12.0.10', [], 0, 0, mktime(0, 59, 0, 7, 4, 2020));
+    $this->assertEquals('13.0.109.0', $json->bundle->version, 'should be offered release');
+
+    // in-schedule, 1 minute after hour
+    $json = $this->getJson('stable', '12.0.10', [], 0, 0, mktime(0, 1, 0, 7, 4, 2020));
+    $this->assertEquals('13.0.109.0', $json->bundle->version, 'should be offered release');
```

Comment:
Shouldn't there be a few tests for 'should not be offered release' - ensuring that the intended delays actually do delay?

Reply:
See line 80, above. We only need to test one item in the schedule because the `ReleaseScheduleTest` class exercises the whole schedule reasonably well.

Reply:
Ah, right, from one of the recent parent PRs.  Forgot about that... and had missed that first test.  :+1: